From 92f605f6bf55bc655d1e4bf91ba3a5c539ef6e07 Mon Sep 17 00:00:00 2001
From: Kevin McCarthy <kevin@8t8.us>
Date: Tue, 3 May 2016 13:21:41 -0700
Subject: [PATCH 29/45] Skip bidi markers in the pager and index. (closes
 #3827)

Curses and slang don't support them, so there's little point in
showing them or attempting to somehow deal with them.

This patch adds filtering in the pager, and changes the filtering
added in 6e0aca94cdb0 for the index to completely skip the marker.
---
 mbyte.c |  8 +++-----
 pager.c | 11 ++++++++++-
 2 files changed, 13 insertions(+), 6 deletions(-)

diff --git a/mbyte.c b/mbyte.c
index fb5718f..41e7f2a 100644
--- a/mbyte.c
+++ b/mbyte.c
@@ -548,14 +548,12 @@ int mutt_filter_unprintable (char **s)
     }
     if (!IsWPrint (wc))
       wc = '?';
-    /* HACK:
-     * Work around a gnu screen bug. See ticket #3827.
-     * Filter out the RIGHT-TO-LEFT and LEFT-TO-RIGHT bidi marks because
-     * they result in screen corruption.
+    /* Filter out the RIGHT-TO-LEFT and LEFT-TO-RIGHT bidi markers because
+     * they result in screen corruption.   See ticket #3827.
      */
     else if (Charset_is_utf8 &&
              ((wc == (wchar_t)0x200f) || (wc == (wchar_t)0x200e)))
-      wc = '?';
+      continue;
     k2 = wcrtomb (scratch, wc, &mbstate2);
     scratch[k2] = '\0';
     mutt_buffer_addstr (b, scratch);
diff --git a/pager.c b/pager.c
index f258aae..24b60ec 100644
--- a/pager.c
+++ b/pager.c
@@ -1154,10 +1154,19 @@ static int format_line (struct line_t **lineInfo, int n, unsigned char *buf,
     if (k == 0)
       k = 1;
 
-    if (Charset_is_utf8 && (wc == 0x200B || wc == 0xFEFF))
+    if (Charset_is_utf8)
     {
+      if (wc == 0x200B || wc == 0xFEFF)
+      {
 	dprint (3, (debugfile, "skip zero-width character U+%04X\n", (unsigned short)wc));
 	continue;
+      }
+      /* Filter bidi markers, see ticket #3827 */
+      if (wc == (wchar_t)0x200f || wc == (wchar_t)0x200e)
+      {
+	dprint (3, (debugfile, "skip bidi marker U+%04X\n", (unsigned short)wc));
+	continue;
+      }
     }
 
     /* Handle backspace */
-- 
2.8.2

